core/src/main/scala/org/locationtech/rasterframes/expressions/package.scala

```diff
@@ -46,6 +46,10 @@ import scala.reflect.runtime.universe._
  * @since 10/10/17
  */
 package object expressions {
+  type HasTernaryExpressionCopy = {def copy(first: Expression, second: Expression, third: Expression): Expression}
+  type HasBinaryExpressionCopy = {def copy(left: Expression, right: Expression): Expression}
+  type HasUnaryExpressionCopy = {def copy(child: Expression): Expression}
```

Comment:
What do you mean by expensive here?

Reply:
Structural types in Scala2 rely on runtime reflection; that's not a big deal but it just got caught by my eye.

Reply:
The better practice is usually to avoid them by having named traits 💭

Reply:
Thank you for calling this out. Can't have named traits because it's trying to call a copy method that will be added by the case class. So it HAS to be runtime reflection. But the whole point of this API change was that runtime reflection was killing Catalyst performance. So while its ugly its better to just go ahead and implement the `withNewChildrenInternal` methods in the case class definitions.